apps/mark-scan/pat-device-input/src/pat_input_reader.rs

```diff
@@ -0,0 +1,231 @@
+                return false;
+            }
+        }
+    }
+
+    pub fn is_signal_b_active(&self) -> bool {
+        match &self.signal_b_pin {
+            Some(pin) => {
+                return pin.is_active();
+            }
+            None => {
+                return false;
+            }
+        }
+    }
+}
+
+pub struct MockPin {
+    address: u16,
+    active: bool,
+    probe_error: Option<io::Error>,
+}
+
+impl Pin for MockPin {
+    fn new(address: u16) -> Self {
+        MockPin {
+            address: address,
+            active: false,
+            probe_error: None,
+        }
+    }
+
+    fn probe(&self) -> Result<(), io::Error> {
+        if let Some(probe_error) = &self.probe_error {
+            // io::Error doesn't implement clone()
+            return Err(io::Error::new(probe_error.kind(), probe_error.to_string()));
+        }
+        Ok(())
+    }
+
+    fn set_up(&self) -> Result<(), io::Error> {
+        Ok(())
+    }
+
+    fn tear_down(&self) {}
+
+    fn is_active(&self) -> bool {
+        self.active
+    }
+}
+
+impl fmt::Display for MockPin {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.address)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use vx_logging::set_app_name;
+
+    use super::*;
+
+    fn get_default_mocked_reader() -> PatInputReader<MockPin> {
+        get_mocked_reader(None)
+    }
+
+    fn get_mocked_reader(probe_err: Option<io::Error>) -> PatInputReader<MockPin> {
+        PatInputReader {
+            is_connected_pin: Some(MockPin {
+                address: IS_CONNECTED_PIN_ADDRESS,
+                active: false,
+                probe_error: probe_err,
+            }),
+            signal_a_pin: Some(MockPin {
+                address: SIGNAL_A_PIN_ADDRESS,
+                active: false,
+                probe_error: None,
+            }),
+            signal_b_pin: Some(MockPin {
+                address: SIGNAL_B_PIN_ADDRESS,
+                active: false,
+                probe_error: None,
+            }),
+        }
+    }
+
+    #[test]
+    fn test_connect_offset() {
+        set_app_name("test");
+        let mut reader = get_default_mocked_reader();
+        reader.connect().unwrap();
+        assert_eq!(
+            reader.is_connected_pin.unwrap().address,
+            IS_CONNECTED_PIN_ADDRESS + POSSIBLE_PIN_OFFSET
+        );
+    }
+
+    #[test]
+    fn test_connect_no_offset() {
+        set_app_name("test");
+        let mut reader =
+            get_mocked_reader(Some(io::Error::new(io::ErrorKind::NotFound, "Test Error")));
+        reader.connect().unwrap();
+        assert_eq!(
+            reader.is_connected_pin.unwrap().address,
+            IS_CONNECTED_PIN_ADDRESS
+        );
+    }
+
+    #[test]
+    fn test_is_connected() {
+        let mut reader = get_default_mocked_reader();
+
+        assert_eq!(reader.is_connected(), false);
+        reader.is_connected_pin.as_mut().unwrap().active = true;
```

Comment:
```suggestion
        assert!(!reader.is_connected());
```

Reply:
I think this suggestion is off by a line but will incorporate manually

Reply:
Grr. Blame VS Code.